js/components/contact_download.jsx

```diff
@@ -0,0 +1,165 @@
+      return;
+    }
+
+    // Pre-fetch the list of agency components, re-purposing the "finder" code.
+    const includeReferenceFields = {
+      agency_component: ['title', 'agency', 'submission_address', 'submission_fax', 'website'],
+      agency: ['name'],
+    };
+    ContactDownloadButton.getFoiaPersonnelFields().forEach((personnelField) => {
+      includeReferenceFields.agency_component.push(personnelField);
+      includeReferenceFields[personnelField] = ['name', 'title', 'email', 'phone'];
+    });
+    requestActions.fetchAgencyFinderData(includeReferenceFields);
+  }
+
+  /**
+   * Download the CSV file.
+   */
+  performDownload() {
+    return ExcellentExport.convert({
+      anchor: this.downloadButton,
+      filename: 'FOIA-Agency-Contacts',
+      format: 'csv',
+    }, [{
+      name: 'FOIA Agency Contacts',
+      from: {
+        array: this.rows,
+      },
+    }]);
+  }
+
+  /**
+   * Event handler for when the download button is triggered.
+   */
+  handleClick() {
+    return (this.rows) ? this.performDownload() : false;
+  }
+
+  /**
+   * Called each time a batch of agency components finished loading.
+   */
+  render() {
+    const progress = agencyComponentStore.getState().agencyFinderDataProgress;
+
+    // If this is not the last batch, just update the button text.
+    if (progress < 100) {
+      if (this.downloadButton) {
+        this.downloadButton.innerHTML = `Please wait - ${progress}%`;
+      }
+    } else {
+      // Otherwise, prepare the CSV data.
+      const { agencyComponents } = this.state;
+
+      // Construct a large array of all the FOIA personnel as rows (objects).
+      const header = [
+        'Component',
+        'Department',
+        'Name',
+        'Title',
+        'Email',
+        'Street Address 1',
+        'Street Address 2',
+        'City',
+        'State',
+        'Zip code',
+        'Telephone',
+        'Fax',
+      ];
+      const rows = [];
+      agencyComponents.forEach((agencyComponent) => {
+        ContactDownloadButton.getFoiaPersonnelFields().forEach((personnelField) => {
+          agencyComponent[personnelField].forEach((foiaPersonnel) => {
+            rows.push([
+              agencyComponent.title,
+              agencyComponent.agency.name,
+              foiaPersonnel.name,
+              foiaPersonnel.title,
+              foiaPersonnel.email,
+              agencyComponent.submission_address.address_line1,
+              agencyComponent.submission_address.address_line2,
+              agencyComponent.submission_address.locality,
+              agencyComponent.submission_address.administrative_area,
+              agencyComponent.submission_address.postal_code,
+              (foiaPersonnel.phone) ? foiaPersonnel.phone.join(', ') : '',
+              agencyComponent.submission_fax,
+            ]);
+          });
+        });
+      });
+
+      // Sort the array of rows by the agency name, then component, then person name.
```

Comment:
Maybe move this sort into `calculateState`, the render function is meant to just take the existing state and turn it into a view. The render function could be called many times, so you're doing an expensive sort on every render, potentially even when the data and rows haven't changed.